### PR TITLE
Add Multicall module

### LINKED ISSUE #2558: Multicall module
This is a very smart piece of code [found in the ENS repository](https://github.com/ensdomains/resolvers/blob/f7d62ab04bfe1692a4344f6f1d31ff81315a98c3/contracts/PublicResolver.sol#L56-L64) that we think could be a useful feature for people to plug into their contracts. I suppose this would be the `utils/Multicall` module, unless we come up with another name!

```solidity
    function multicall(bytes[] calldata data) external returns(bytes[] memory results) {
        results = new bytes[](data.length);
        for(uint i = 0; i < data.length; i++) {
            (bool success, bytes memory result) = address(this).delegatecall(data[i]);
            require(success);
            results[i] = result;
        }
        return results;
    }
```

### PR DESCRIPTION
**Pull request by martriay**
Fixes #2558 

- [x] Tests
- [x] Documentation
- [x] Changelog entry

### CODE REVIEW

**Review by Amxx — COMMENTED** [review]
Code looks good to me. 
You could shorten it even more by doing :
```results[i] = Address.functionDelegateCall(address(this), data[i]);```

Depending on how solidity does optimization, that might even save a memory copy.
Next we'll need tests and documentation

**Comment by Amxx** [comment]
For the record: https://github.com/Uniswap/uniswap-v3-periphery/blob/main/contracts/base/Multicall.sol
It would be great to measure gas usage against this

**Comment by martriay** [comment]
Doing some quick tests I see this

```
·------------------------------------|----------------------------|-------------|-----------------------------·
|        Solc version: 0.8.0         ·  Optimizer enabled: false  ·  Runs: 200  ·  Block limit: 10000000 gas  │
·····································|····························|·············|······························
|  Methods                                                                                                    │
·······················|·············|··············|·············|·············|···············|··············
|  Contract            ·  Method     ·  Min         ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
·······················|·············|··············|·············|·············|···············|··············
|  BatchCallTokenMock  ·  batchCall  ·           -  ·          -  ·      88620  ·            1  ·          -  │
·······················|·············|··············|·············|·············|···············|··············
|  Deployments                       ·                                          ·  % of limit   ·             │
·····································|··············|·············|·············|···············|··············
|  BatchCallTokenMock                ·           -  ·          -  ·    1897353  ·         19 %  ·          -  │
·------------------------------------|--------------|-------------|-------------|---------------|-------------·
```

```
·-----------------------------|----------------------------|-------------|-----------------------------·
|     Solc version: 0.8.0     ·  Optimizer enabled: false  ·  Runs: 200  ·  Block limit: 10000000 gas  │
······························|····························|·············|······························
|  Methods                                                                                             │
················|·············|··············|·············|·············|···············|··············
|  Contract     ·  Method     ·  Min         ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
················|·············|··············|·············|·············|···············|··············
|  Deployments                ·                                          ·  % of limit   ·             │
······························|··············|·············|·············|···············|··············
|  UniswapMultiCallTokenMock  ·           -  ·          -  ·    1918374  ·       19.2 %  ·          -  │
·-----------------------------|--------------|-------------|-------------|---------------|-------------·
```

**Comment by martriay** [comment]
```solidity
// SPDX-License-Identifier: GPL-2.0-or-later
pragma solidity 0.8.0;
pragma abicoder v2;

import "./ERC20Mock.sol";

/// @title Multicall
/// @notice Enables calling multiple methods in a single call to the contract
abstract contract Multicall {
    function multicall(bytes[] calldata data) external payable returns (bytes[] memory results) {
        results = new bytes[](data.length);
        for (uint256 i = 0; i < data.length; i++) {
            (bool success, bytes memory result) = address(this).delegatecall(data[i]);

            if (!success) {
                // Next 5 lines from https://ethereum.stackexchange.com/a/83577
                if (result.length < 68) revert();
                assembly {
                    result := add(result, 0x04)
                }
                revert(abi.decode(result, (string)));
            }

            results[i] = result;
        }
    }
}

contract UniswapMultiCallTokenMock is ERC20Mock, Multicall {
    constructor (uint256 initialBalance) ERC20Mock("BatchCallToken", "BCT", msg.sender, initialBalance) {}
}
```

v.s.

```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

import "../utils/BatchCall.sol";
import "./ERC20Mock.sol";

contract BatchCallTokenMock is ERC20Mock, BatchCall {
    constructor (uint256 initialBalance) ERC20Mock("BatchCallToken", "BCT", msg.sender, initialBalance) {}
}
```

**Comment by martriay** [comment]
Doing some quick tests:


```solidity
// SPDX-License-Identifier: GPL-2.0-or-later
pragma solidity 0.8.0;
pragma abicoder v2;

import "./ERC20Mock.sol";

/// @title Multicall
/// @notice Enables calling multiple methods in a single call to the contract
abstract contract Multicall {
    function multicall(bytes[] calldata data) external payable returns (bytes[] memory results) {
        results = new bytes[](data.length);
        for (uint256 i = 0; i < data.length; i++) {
            (bool success, bytes memory result) = address(this).delegatecall(data[i]);

            if (!success) {
                // Next 5 lines from https://ethereum.stackexchange.com/a/83577
                if (result.length < 68) revert();
                assembly {
                    result := add(result, 0x04)
                }
                revert(abi.decode(result, (string)));
            }

            results[i] = result;
        }
    }
}

contract UniswapMultiCallTokenMock is ERC20Mock, Multicall {
    constructor (uint256 initialBalance) ERC20Mock("BatchCallToken", "BCT", msg.sender, initialBalance) {}
}
```

v.s.

```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

import "../utils/BatchCall.sol";
import "./ERC20Mock.sol";

contract BatchCallTokenMock is ERC20Mock, BatchCall {
    constructor (uint256 initialBalance) ERC20Mock("BatchCallToken", "BCT", msg.sender, initialBalance) {}
}
```

results in

```
·------------------------------------|----------------------------|-------------|-----------------------------·
|        Solc version: 0.8.0         ·  Optimizer enabled: false  ·  Runs: 200  ·  Block limit: 10000000 gas  │
·····································|····························|·············|······························
|  Methods                                                                                                    │
·······················|·············|··············|·············|·············|···············|··············
|  Contract            ·  Method     ·  Min         ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
·······················|·············|··············|·············|·············|···············|··············
|  BatchCallTokenMock  ·  batchCall  ·           -  ·          -  ·      88620  ·            1  ·          -  │
·······················|·············|··············|·············|·············|···············|··············
|  Deployments                       ·                                          ·  % of limit   ·             │
·····································|··············|·············|·············|···············|··············
|  BatchCallTokenMock                ·           -  ·          -  ·    1897353  ·         19 %  ·          -  │
·------------------------------------|--------------|-------------|-------------|---------------|-------------·
```

```
·-------------------------------------------|----------------------------|-------------|-----------------------------·
|            Solc version: 0.8.0            ·  Optimizer enabled: false  ·  Runs: 200  ·  Block limit: 10000000 gas  │
············································|····························|·············|······························
|  Methods                                                                                                           │
······························|·············|··············|·············|·············|···············|··············
|  Contract                   ·  Method     ·  Min         ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
······························|·············|··············|·············|·············|···············|··············
|  UniswapMultiCallTokenMock  ·  multicall  ·           -  ·          -  ·      85551  ·            1  ·          -  │
······························|·············|··············|·············|·············|···············|··············
|  Deployments                              ·                                          ·  % of limit   ·             │
············································|··············|·············|·············|···············|··············
|  UniswapMultiCallTokenMock                ·           -  ·          -  ·    1918374  ·       19.2 %  ·          -  │
·-------------------------------------------|--------------|-------------|-------------|---------------|-------------·
```

**Comment by Amxx** [comment]
> ```
> ·------------------------------------|----------------------------|-------------|-----------------------------·
> |        Solc version: 0.8.0         ·  Optimizer enabled: false  ·  Runs: 200  ·  Block limit: 10000000 gas  │
> ·····································|····························|·············|······························
> |  Methods                                                                                                    │
> ·······················|·············|··············|·············|·············|···············|··············
> |  Contract            ·  Method     ·  Min         ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
> ·······················|·············|··············|·············|·············|···············|··············
> |  BatchCallTokenMock  ·  batchCall  ·           -  ·          -  ·      88620  ·            1  ·          -  │
> ·······················|·············|··············|·············|·············|···············|··············
> |  Deployments                       ·                                          ·  % of limit   ·             │
> ·····································|··············|·············|·············|···············|··············
> |  BatchCallTokenMock                ·           -  ·          -  ·    1897353  ·         19 %  ·          -  │
> ·------------------------------------|--------------|-------------|-------------|---------------|-------------·
> ```
> 
> 
> ```
> ·-------------------------------------------|----------------------------|-------------|-----------------------------·
> |            Solc version: 0.8.0            ·  Optimizer enabled: false  ·  Runs: 200  ·  Block limit: 10000000 gas  │
> ············································|····························|·············|······························
> |  Methods                                                                                                           │
> ······························|·············|··············|·············|·············|···············|··············
> |  Contract                   ·  Method     ·  Min         ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
> ······························|·············|··············|·············|·············|···············|··············
> |  UniswapMultiCallTokenMock  ·  multicall  ·           -  ·          -  ·      85551  ·            1  ·          -  │
> ······························|·············|··············|·············|·············|···············|··············
> |  Deployments                              ·                                          ·  % of limit   ·             │
> ············································|··············|·············|·············|···············|··············
> |  UniswapMultiCallTokenMock                ·           -  ·          -  ·    1918374  ·       19.2 %  ·          -  │
> ·-------------------------------------------|--------------|-------------|-------------|---------------|-------------·
> ```

This shows that, without optimisation, uniswap multicall is more expensive to deploy by about 20k gas, but less expensive to use..
We would need to run that with optimisation though ...

File: contracts/utils/BatchCall.sol [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "./Address.sol";
+
+/*
```
Comment (frangio) [inline contracts/utils/BatchCall.sol:7]:
```suggestion
/**
```

**Comment by martriay** [comment]
Should I switch to their implementation?

```
·-------------------------------------------|---------------------------|-------------|-----------------------------·
|            Solc version: 0.8.0            ·  Optimizer enabled: true  ·  Runs: 200  ·  Block limit: 10000000 gas  │
············································|···························|·············|······························
|  Methods                                                                                                          │
······························|·············|·············|·············|·············|···············|··············
|  Contract                   ·  Method     ·  Min        ·  Max        ·  Avg        ·  # calls      ·  usd (avg)  │
······························|·············|·············|·············|·············|···············|··············
|  BatchCallTokenMock         ·  batchCall  ·          -  ·          -  ·      85618  ·            1  ·          -  │
······························|·············|·············|·············|·············|···············|··············
|  UniswapBatchCallTokenMock  ·  multicall  ·          -  ·          -  ·      82695  ·            1  ·          -  │
······························|·············|·············|·············|·············|···············|··············
|  Deployments                              ·                                         ·  % of limit   ·             │
············································|·············|·············|·············|···············|··············
|  BatchCallTokenMock                       ·          -  ·          -  ·    1139352  ·       11.4 %  ·          -  │
············································|·············|·············|·············|···············|··············
|  UniswapBatchCallTokenMock                ·          -  ·          -  ·    1120295  ·       11.2 %  ·          -  │
·-------------------------------------------|-------------|-------------|-------------|---------------|-------------·
```

**Comment by frangio** [comment]
The additional cost is probably that we are calling `extcodesize` in `functionDelegateCall`. I think we should keep this check, and it will become a lot cheaper with the Berlin hardfork anyway.

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,7 @@
  * `IERC20Metadata`: add a new extended interface that includes the optional `name()`, `symbol()` and `decimals()` functions. ([#2561](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2561))
  * `ERC777`: make reception acquirement optional in `_mint`. ([#2552](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2552))
  * `ERC20Permit`: add a `_useNonce` to enable further usage of ERC712 signatures. ([#2565](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2565))
+ * `Multicall`: add abstract contract with `Multicall(bytes[] calldata data)` function to bundle multiple calls together ([#2608](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2608))
```
Comment (frangio) [inline CHANGELOG.md:8]:
```suggestion
 * `Multicall`: add abstract contract with `multicall(bytes[] calldata data)` function to bundle multiple calls together ([#2608](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2608))
```

File: docs/modules/ROOT/pages/utilities.adoc [hunk 3]
```diff
@@ -99,3 +99,41 @@ Want to check if an address is a contract? Use xref:api:utils.adoc#Address[`Addr
...
+        ...
+    }
+
+    function bar() public {
+        ...
+    }
+}
+----
+
+This is how to call the `multicall` function using Truffle, allowing `foo` and `bar` to be called in a single transaction:
```
Comment (frangio) [inline docs/modules/ROOT/pages/utilities.adoc:127]: Awesome to have this example in the docs!

File: test/utils/BatchCall.test.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+const { BN, expectRevert } = require('@openzeppelin/test-helpers');
+const MulticallTokenMock = artifacts.require('MulticallTokenMock');
+
+contract('MulticallToken', function (accounts) {
```
Comment (frangio) [inline test/utils/BatchCall.test.js:4]: We're missing two important tests:

- A revert in one of the subcalls will revert all of the state changes of previous subcalls.
- The return value is an array of the return values of subcalls, in the order they were specified.

**Review by frangio — COMMENTED** [review]
Other than my suggestions for testing this looks good.

**Review by Amxx — DISMISSED** [review]
All this looks good to me, I'd personally go with uniswap implementation if that can save some gas, or do a manual try-catch block ... but that is not a strong opinion.

**Comment by martriay** [comment]
Linking @Amxx's [`payable` exploit analysis]( https://github.com/Uniswap/uniswap-v3-periphery/issues/52) here as rationale for removing the `payable` modifier.

**Review by frangio — APPROVED** [review]
Thank you all!

**Pull request merged**